### PR TITLE
Add optuna option for searchers

### PR DESCRIPTION
**Pull request by fraimondo**
Integrate Optuna (https://optuna.org/) for Hyperparameter Tuning as `optuna` searcher, using the `OptunaSearchCV` model.

### CODE REVIEW

File: julearn/model_selection/tests/test_skopt_searcher.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
+"""Provides tests for the bayes searcher."""
+
+# Authors: Federico Raimondo <f.raimondo@fz-juelich.de>
+# License: AGPL
+from typing import Dict
```
Comment (synchon) [inline julearn/model_selection/tests/test_skopt_searcher.py:5]: Newline above this.

File: julearn/model_selection/tests/test_skopt_searcher.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                "more_options": sksp.Categorical(
+                    ("a", "b", "c", "d"),
+                ),
+            },
+            (None, None),
+            ("categorical", "categorical"),
+        ),
+    ],
+)
+def test__prepare_skopt_hyperparameters_distributions(
```
Comment (synchon) [inline julearn/model_selection/tests/test_skopt_searcher.py:71]: The typing annotations need to be adjusted similar to my comments in test_optuna_searcher.py

File: julearn/pipeline/pipeline_creator.py [hunk 3]
```diff
@@ -928,8 +934,8 @@ def _prepare_hyperparameter_tuning(
         The parameters for the search. The following keys are accepted:
 
         * 'kind': The kind of search algorithm to use e.g.:
-            'grid', 'random' or 'bayes'. All valid julearn searchers can be
-            entered.
+            'grid', 'random', 'bayes' or 'optuna'. All valid julearn searchers
```
Comment (synchon) [inline julearn/pipeline/pipeline_creator.py:937]: Might be worth it to make a listing.

**Review by synchon — CHANGES_REQUESTED** [review]

**Comment by synchon** [comment]
The PR also needs a description for future reference.

File: docs/changes/newsfragments/262.enh [hunk 4]
```diff
@@ -0,0 +1 @@
+Add :class:`~optuna_integration.sklearn.OptunaSearchCV` to the list of available searchers as 'optuna', update documentation on Hyperparameter tuning and refactor how hyperparmeters' distributions are specified by `Fede Raimondo`_
```
[the remark this answers is not in the document]
Comment (fraimondo) [reply docs/changes/newsfragments/262.enh:1]: Which categories should be?

File: julearn/api.py [hunk 5]
```diff
@@ -136,8 +136,8 @@ def run_cross_validation(  # noqa: C901
         the following keys:
 
         * 'kind': The kind of search algorithm to use, e.g.:
-            'grid', 'random' or 'bayes'. Can be any valid julearn searcher name
-            or scikit-learn compatible searcher.
+            'grid', 'random', 'bayes' or 'optuna'. Can be any valid julearn
```
[the remark this answers is not in the document]
Comment (fraimondo) [reply julearn/api.py:139]: The user can also "register" a searcher.

[the remark this answers is not in the document]
Comment (synchon) [reply julearn/api.py:139]: I'd do something like:

```
* ``"grid"`` : <link to the class>
* ``"random"`` : <link to the class>
* ``"bayes"`` : <link to the class>
* ``"optuna"`` : <link to the class>
* valid julearn searcher name (<link to API ref if possible>)
* ``scikit-learn``-compatible searcher
```

File: docs/changes/newsfragments/262.enh (hunk 4 above)
[the remark this answers is not in the document]
Comment (synchon) [reply docs/changes/newsfragments/262.enh:1]: 262.feature :
```
Add :class:`~optuna_integration.sklearn.OptunaSearchCV` to the list of available searchers as ``"optuna"`` ...
```

262.doc :
```
Update documentation on Hyperparameter tuning ...
```

262.enh :
```
Refactor how hyperparmeters' distributions are specified ...
```

File: julearn/model_selection/tests/test_available_searchers.py [hunk 6]
```diff
@@ -62,17 +62,32 @@ def test_get_searcher() -> None:
     out = get_searcher("bayes")
     assert out.__name__ == "BayesSearchCV"
 
+    out = get_searcher("optuna")
```
Comment (synchon) [inline julearn/model_selection/tests/test_available_searchers.py:65]: Missed it earlier, might be worth it to adapt this to use parametrization.

File: julearn/model_selection/tests/test_optuna_searcher.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            },
+            (None, None),
+            ("categorical", "categorical"),
+        ),
+    ],
+)
+def test__prepare_optuna_hyperparameters_distributions(
+    params_to_tune: Dict[str, Dict[str, Any]],
+    expected_types: Tuple,
+    expected_dist: tuple,
```
Comment (synchon) [inline julearn/model_selection/tests/test_optuna_searcher.py:75]: tuple -> Tuple (typing.Tuple) [to support Python 3.8]

**Review by synchon — APPROVED** [review]

**Pull request merged**